camera.c

```diff
@@ -0,0 +1,298 @@
+	for (int i = 0, j = 0; i < frame_size; i += 4, j += 6)
+	{
+		int Y0 = yuyv[i];
+		int U = yuyv[i + 1] - 128;
+		int Y1 = yuyv[i + 2];
+		int V = yuyv[i + 3] - 128;
+
+		int R1 = Y0 + 1.140 * V;
+		int G1 = Y0 - 0.395 * U - 0.581 * V;
+		int B1 = Y0 + 2.032 * U;
+
+		int R2 = Y1 + 1.140 * V;
+		int G2 = Y1 - 0.395 * U - 0.581 * V;
+		int B2 = Y1 + 2.032 * U;
+
+		rgb_ref[j + 0] = (uint8_t)(R1 < 0 ? 0 : (R1 > 255 ? 255 : R1));
+		rgb_ref[j + 1] = (uint8_t)(G1 < 0 ? 0 : (G1 > 255 ? 255 : G1));
+		rgb_ref[j + 2] = (uint8_t)(B1 < 0 ? 0 : (B1 > 255 ? 255 : B1));
+		rgb_ref[j + 3] = (uint8_t)(R2 < 0 ? 0 : (R2 > 255 ? 255 : R2));
+		rgb_ref[j + 4] = (uint8_t)(G2 < 0 ? 0 : (G2 > 255 ? 255 : G2));
+		rgb_ref[j + 5] = (uint8_t)(B2 < 0 ? 0 : (B2 > 255 ? 255 : B2));
+	}
+}
+
+void rgb_to_jpeg(uint8_t **jpeg_ref, unsigned long *size_ref, const uint8_t *rgb, const int width, const int height)
+{
+	struct jpeg_compress_struct cinfo;
+	struct jpeg_error_mgr jerr;
+
+	JSAMPROW row_pointer[1];
+	int row_stride;
+
+	cinfo.err = jpeg_std_error(&jerr);
+	jpeg_create_compress(&cinfo);
+
+	jpeg_mem_dest(&cinfo, jpeg_ref, size_ref);
+
+	cinfo.image_width = width;
+	cinfo.image_height = height;
+	cinfo.input_components = 3;
+	cinfo.in_color_space = JCS_RGB;
+
+	jpeg_set_defaults(&cinfo);
+	jpeg_set_quality(&cinfo, 75, TRUE);
+
+	jpeg_start_compress(&cinfo, TRUE);
+
+	row_stride = width * 3;
+
+	while (cinfo.next_scanline < cinfo.image_height)
+	{
+		row_pointer[0] = (JSAMPROW)&rgb[cinfo.next_scanline * row_stride];
+		jpeg_write_scanlines(&cinfo, row_pointer, 1);
+	}
+
+	jpeg_finish_compress(&cinfo);
+	jpeg_destroy_compress(&cinfo);
+}
+
+int camera_capture_jpeg(uint8_t **buffer_ref, unsigned long *size_ref, const char *video_device)
+{
+	int fd = -1;
+	uint8_t *raw_buffer = NULL;
+	int raw_size = 0;
+
+	void cleanup()
+	{
+		if (fd >= 0)
+		{
+			close(fd);
+		}
+		if (raw_buffer != NULL)
+		{
+			munmap(raw_buffer, raw_size);
+		}
+	}
+
+	int ret = 0;
+	char *driver_str;
+
+	fprintf(stderr, "Opening device: %s\n", video_device);
+	ret = init_device(video_device, &fd, &driver_str);
+	if (ret < 0)
+	{
+		cleanup();
+		return -1;
+	}
+
+	int width, height;
+	uint32_t format;
+	ret = configure_format(fd, &width, &height, &format);
+	if (ret < 0)
+	{
+		cleanup();
+		return -1;
+	}
+
+	fprintf(stderr, "Driver: %s, Resolution: %dx%d, Format: %c%c%c%c\n", driver_str, width, height,
+		(char)((format >> 0) & 0xFF), (char)((format >> 8) & 0xFF), (char)((format >> 16) & 0xFF), (char)((format >> 24) & 0xFF));
+	free(driver_str);
+
+	ret = request_buffer(fd);
+	if (ret < 0)
+	{
+		cleanup();
+		return -1;
+	}
+
+	struct v4l2_buffer buf_info = {0};
+
+	ret = query_buffer(fd, &raw_buffer, &buf_info, &raw_size);
+	if (ret < 0)
+	{
+		cleanup();
+		return -1;
+	}
+
+	capture_frame(fd, &buf_info);
+
+	uint8_t *jpeg = NULL;
+	switch (format)
+	{
+		case V4L2_PIX_FMT_YUYV:
+		{
+			uint8_t *rgb = (uint8_t *)malloc(width * height * 3);
+
+			yuyv_to_rgb(rgb, raw_buffer, width, height);
```

Comment:
Why do you need to convert YUYV to RGB? JPEG compression will convert the image back to YCbCr. Can we use JCS_YCbCr instead? If there is a need for pixel repacking of planar conversion, it will still be easier to handle than doing YUV -> RGB -> YUV.